File: grind75/Dec20/542.py
from collections import deque
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        # bfs
        q = deque()
        m = len(mat)
        n = len(mat[0])
        directions = [[1,0], [-1,0], [0,1], [0,-1]]
        for i in range(m):
            for j in range(n):
                if mat[i][j] == 0:
                    q.appendleft((i, j, 0))
                else:
                    mat[i][j] = math.inf
        
        while q:
            i,j,d = q.pop()
            if mat[i][j] > d:
                mat[i][j] = d
            
            for direction in directions:
                new_i = i + direction[0]
                new_j = j + direction[1]
                if new_i >= 0 and new_i < m and new_j >= 0 and new_j < n 
and mat[new_i][new_j] > (d+1):
                    q.appendleft((new_i,new_j,d+1))
            
        return mat

File: grind75/Dec20/542.go
func updateMatrix(mat [][]int) [][]int {
    dirs := [][]int{[]int{0,1}, []int{0,-1}, []int{1,0}, []int{-1,0},}
    m,n := len(mat), len(mat[0])
    visited := make([]int, m*n)
    q := [][]int{}
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            if mat[i][j] == 0 {
                q = append(q, []int{i,j})
                visited[i*n+j] = 1
            }
        }
    }
    dist := 0
    for len(q) > 0 {
        cq := [][]int{}
        lq := len(q)
        dist += 1
        for lq > 0 {
            node := q[0]
            q = q[1:]
            lq--
            i,j := node[0], node[1]
            for _, dir := range dirs {
                ni,nj := i+dir[0], j+dir[1]
                if ni < 0 || ni >= m || nj < 0 || nj >= n { continue }
                if visited[ni*n+nj] == 1 { continue }
                visited[ni*n+nj] = 1
                mat[ni][nj] = dist
                cq = append(cq, []int{ni,nj})
            }
        }
        q = cq
    }
    return mat
}
